subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -42,7 +43,8 @@
     private final Map<String, ImmutableList<Path>> watchedRootsForSnapshot = new HashMap<>();
     private final Multiset<Path> shouldWatchDirectories = HashMultiset.create();
     private final Set<Path> watchedRoots = new HashSet<>();
-    private final Set<Path> mustWatchDirectories = new HashSet<>();
+    private final List<Path> projectRootDirectories = new ArrayList<>();
+    private final List<String> projectRootDirectoryPrefixes = new ArrayList<>();
```

Comment:
Assuming a single root directory could allow simplifying the code here significantly.

Reply:
Done.